File: code/C#/Typeclasses/FBounded/Ord.cs
namespace Typeclasses.FBounded
{
    public interface Ord<in T>
    {
        bool gt(T that);
        bool eq(T that);
    }

    public class IntOrd : Ord<IntOrd>
    {
        public int Value { get; set; }

        public bool gt(IntOrd that) => this.Value > that.Value;
        public bool eq(IntOrd that) => this.Value == that.Value;
    }
    
    public class StringOrd : Ord<StringOrd>
    {
        public string Value { get; set; }

        public bool gt(StringOrd that) => string.CompareOrdinal(this.Value, that.Value) > 0;
        public bool eq(StringOrd that) => string.CompareOrdinal(this.Value, that.Value) == 0;
    }
}

File: code/Scala/src/main/scala/net/sigusr/implicit1/Ord.scala
package net.sigusr.implicit1

sealed trait Ord[-T] {
  def gt(l: T, r: T): Boolean
  def eq(l: T, r: T): Boolean
}

object instances {
  implicit val intOrd: Ord[Int] = new Ord[Int] {
    override def gt(l: Int, r: Int): Boolean = l > r
    override def eq(l: Int, r: Int): Boolean = l == r
  }

  implicit val stringOrd: Ord[String] = new Ord[String] {
    override def gt(l: String, r: String): Boolean = l > r
    override def eq(l: String, r: String): Boolean = l == r
  } 
}

